pandas/core/arrays/categorical.py

```diff
@@ -2717,6 +2718,15 @@ def _groupby_op(
             res_values[result_mask == 1] = -1
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other(
+        self, other: Categorical, sort: bool = False  # type: ignore[override]
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        other = self._encode_with_my_categories(other)
+
+        lk = ensure_int64(self.codes)
+        rk = ensure_int64(other.codes)
```

Comment:
is the ensure_int64 necessary?  im betting that is from a time before we had non-64 hashtables

Reply:
we run into trouble if they have different int dtypes, so casting both to int64 seems sensible

Reply:
they should be matching dtypes though, right?

Reply:
simple reproducer:

```
ser = pd.Series([1, 2, 3, 1, 2], dtype="category")
cat = pd.Series(pd.Categorical.from_codes(ser.cat.codes.astype("int32"), ser.cat.categories))
assert ser.dtype == cat.dtype
```

dtype doesn't take the codes dtype into account

Reply:
We can also make this more efficient and do more extensive checking, but I think that we need something like this

Reply:
isnt that what encode_with_my_categories is for?  and even before that is called shouldn't we only get here with self.dtype == other.dtype?

Reply:
The dtypes are equal, the assertion passes.

```
_encode_with_my_categories
```
basically returns the same codes if the categories are equal

Reply:
i maintain the ensure_int64 shouldn't be necessary.  if you strongly disagree, can you add a TODO to follow up, pointing back to the relevant part of this thread

Reply:
Sorry you are correct, removed the cast